wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -0,0 +1,2592 @@
+				$this->execute_update_statement( $ast );
+				break;
+			case 'replaceStatement':
+				$this->query_type = 'REPLACE';
+				$this->execute_insert_or_replace_statement( $ast );
+				break;
+			case 'deleteStatement':
+				$this->query_type = 'DELETE';
+				$this->execute_delete_statement( $ast );
+				break;
+			case 'createStatement':
+				$this->query_type = 'CREATE';
+				$subtree          = $ast->get_child_node();
+				switch ( $subtree->rule_name ) {
+					case 'createTable':
+						$this->execute_create_table_statement( $ast );
+						break;
+					default:
+						throw new Exception(
+							sprintf(
+								'Unsupported statement type: "%s" > "%s"',
+								$ast->rule_name,
+								$subtree->rule_name
+							)
+						);
+				}
+				break;
+			case 'alterStatement':
+				$this->query_type = 'ALTER';
+				$subtree          = $ast->get_child_node();
+				switch ( $subtree->rule_name ) {
+					case 'alterTable':
+						$this->execute_alter_table_statement( $ast );
+						break;
+					default:
+						throw new Exception(
+							sprintf(
+								'Unsupported statement type: "%s" > "%s"',
+								$ast->rule_name,
+								$subtree->rule_name
+							)
+						);
+				}
+				break;
+			case 'dropStatement':
+				$this->query_type = 'DROP';
+				$subtree          = $ast->get_child_node();
+				switch ( $subtree->rule_name ) {
+					case 'dropTable':
+						$this->execute_drop_table_statement( $ast );
+						break;
+					default:
+						$query = $this->translate( $ast );
+						$this->execute_sqlite_query( $query );
+						$this->set_result_from_affected_rows();
+				}
+				break;
+			case 'setStatement':
+				/*
+				 * It would be lovely to support at least SET autocommit,
+				 * but I don't think that is even possible with SQLite.
+				 */
+				$this->results = 0;
+				break;
+			case 'showStatement':
+				$this->query_type = 'SHOW';
+				$this->execute_show_statement( $ast );
+				break;
+			case 'utilityStatement':
+				$this->query_type = 'DESCRIBE';
+				$subtree          = $ast->get_child_node();
+				switch ( $subtree->rule_name ) {
+					case 'describeStatement':
+						$this->execute_describe_statement( $subtree );
+						break;
+					default:
+						throw new Exception(
+							sprintf(
+								'Unsupported statement type: "%s" > "%s"',
+								$ast->rule_name,
+								$subtree->rule_name
+							)
+						);
+				}
+				break;
+			default:
+				throw new Exception( sprintf( 'Unsupported statement type: "%s"', $ast->rule_name ) );
+		}
+	}
+
+	private function execute_select_statement( WP_Parser_Node $node ): void {
+		$has_sql_calc_found_rows = null !== $node->get_descendant_token(
+			WP_MySQL_Lexer::SQL_CALC_FOUND_ROWS_SYMBOL
+		);
+
+		// First, translate the query, before we modify last found rows count.
+		$query = $this->translate( $node->get_child() );
+
+		// Handle SQL_CALC_FOUND_ROWS.
+		if ( true === $has_sql_calc_found_rows ) {
+			// Recursively find a query expression with the first LIMIT or SELECT.
```

Comment:
this was almost impossible to get right in the old driver and now we can just reason about a tree – this is fantastic